### PR TITLE
PrettyPrinterPerformance Optimized the PrettyPrinter for #894

### LINKED ISSUE #894: #883 regressed performance by ~7%
https://github.com/swiftlang/swift-format/pull/883 regressed swift-format’s performance by 7-8%. To reproduce run

```bash
cd /tmp
git clone https://github.com/apple/swift-syntax.git
cd swift-syntax
git checkout -f 509.0.0

cd path/to/swift-format
swift build -c release
.build/release/swift-format lint --measure-instructions --recursive /tmp/swift-syntax
```

And compare the instruction count between swift-format with and without https://github.com/swiftlang/swift-format/pull/883.

### PR DESCRIPTION
**Pull request by macshome**
Worked to get the perfomance to be closer to where we were before the changes in #883. This code should be only about 1.5% slower rather than 7% slower.

Pre-changes:  Instructions executed: 67167400341
After #883: Instructions executed: 71260352299
This PR: Instructions executed: 68259258073

Fixes #894.

### CODE REVIEW

**Review by ahoppen — APPROVED** [review]

**Comment by ahoppen** [comment]
Thank you, @macshome, that's a great improvement.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
Looks like this fails to build using Swift 5.10. Maybe the `where` label for `count(where:)` is required? Could you check the failure?

**Comment by macshome** [comment]
> Unfortunately, `count(where:)` didn't actually land until Swift 6.0: https://github.com/swiftlang/swift-evolution/blob/main/proposals/0220-count-where.md
> 
> It's probably fine here to just write the loop manually instead of using that API. If you simultaneously keep track of the latest `\n` index every time through the loop, then you can do everything in a single forward scan and avoid a separate call to `lastIndex` afterwards, which should shave off a few more instructions. Every bit helps in this critical path.

Will do. This explains why it was failing on Godbolt when including `count(where:_)` for me as well.

**Comment by macshome** [comment]
I find it interesting that the `text.range(of: "\n", options: .backwards)` option has less instructions executed, even though it actually has 3X more assembly instructions than the `lastNewlineIndex = text.lastIndex(of: "\n")` version.

https://godbolt.org/z/v9xjnq4qb

I guess that the benefit of reading the text corpus backwards makes up for it.

**Comment by allevato** [comment]
`lastIndex` should also be reading the string backwards, since `String` conforms to `BidirectionalCollection`.

When you're looking at the instruction counts, are you getting those by running swift-format on a Mac or a Linux machine? Godbolt.org only has Linux machines AFAIK (at least for Swift), so if you're looking at Godbolt output there, you could be comparing something vastly different than what you would see on macOS. And even on recent versions of macOS, they've been porting Foundation from the old Obj-C implementation to the new Swift core, so the code you're executing could even vary significantly between versions there, too.

Ultimately, I think this is still doing unnecessary work (either approach); in my original comment, I suggested having a _single loop_ that counts the number of `"\n"`s and simultaneously keeps track of the index of the last one that it's seen, rather than doing a separate counting operation and then a backwards scan. This should be entirely possible to do in just one sequential pass over the string, which should it get it closest to the original performance.

**Comment by allevato** [comment]
Another optimization would be to iterate over the `.utf8` view of the string instead of the string directly. Since all we care about here is finding the number and locations of the `"\n"` characters, we don't need to pay the cost of any of the grapheme clustering that comes along with `Character`-based APIs.

**Comment by macshome** [comment]
> `lastIndex` should also be reading the string backwards, since `String` conforms to `BidirectionalCollection`.
> 
> When you're looking at the instruction counts, are you getting those by running swift-format on a Mac or a Linux machine? Godbolt.org only has Linux machines AFAIK (at least for Swift), so if you're looking at Godbolt output there, you could be comparing something vastly different than what you would see on macOS. And even on recent versions of macOS, they've been porting Foundation from the old Obj-C implementation to the new Swift core, so the code you're executing could even vary significantly between versions there, too.
> 
> Ultimately, I think this is still doing unnecessary work (either approach); in my original comment, I suggested having a _single loop_ that counts the number of `"\n"`s and simultaneously keeps track of the index of the last one that it's seen, rather than doing a separate counting operation and then a backwards scan. This should be entirely possible to do in just one sequential pass over the string, which should it get it closest to the original performance.

Good point. I'll check the assembly on my Mac as well when working on additional optimizations.

Thanks!

**Comment by macshome** [comment]
I do have a question that isn't about the optimization work: should we be counting newlines or lines?

This string:
```swift
let text = """
This is some text
it has a lot 
of
line

breaks.
```
has 5 `"\n"` but 6 lines. In general there is always one more line than there is line breaks. If I try adding an additional line to the count of `"\n"` a bunch of tests will fail.

Is the code handling this conversion of newlines to actual lines somewhere else so that we don't need to worry about it here?

**Comment by macshome** [comment]
I've pretty much got this down to one forward pass that equals the original instruction count now. Just adding in the last tweaks.

**Comment by macshome** [comment]
@ahoppen @allevato 

Between making this one forward loop, and using the `.utf8` view, the benchmark now reports that it is faster than it was pre-#883.

Instructions executed: 66824319169

**Review by ahoppen — COMMENTED** [review]
Amazing work for beating the original baseline! 🚀

File: Sources/SwiftFormat/PrettyPrint/PrettyPrintBuffer.swift [hunk 1]
```diff
@@ -122,16 +122,17 @@ struct PrettyPrintBuffer {
...
-      // in case we have inserted new lines, we need to reset the column
-      column = lines.last?.count ?? 0
-    } else {
-      // in case it is an end of line comment or a single line comment,
-      // we just add to the current column
-      column += lines.last?.count ?? 0
+    var lastLength = 0
+    // We are only interested in "\n" we can use the UTF8 view and skip the grapheme clustering.
+    for element in text.utf8 {
+      if element == 10 {
```
[the remark this answers is not in the document]
Comment (macshome) [reply Sources/SwiftFormat/PrettyPrint/PrettyPrintBuffer.swift]: Done! I couldn't find the spelling of that to save my life. Thanks.

**Review by ahoppen — APPROVED** [review]

**Pull request merged**